typing/typecore.ml

```diff
@@ -1545,14 +1546,56 @@ and type_pat_aux
         raise(Error(loc, !env, Constructor_arity_mismatch(lid.txt,
                                      constr.cstr_arity, List.length sargs)));
       begin_def ();
-      let (ty_args, ty_res) =
-        instance_constructor ~in_pattern:(env, get_gadt_equations_level ())
-          constr
-      in
       let expected_ty = instance expected_ty in
       (* PR#7214: do not use gadt unification for toplevel lets *)
-      unify_pat_types loc env ty_res expected_ty
-        ~refine:(refine || constr.cstr_generalized && no_existentials = None);
+      let unify_res ty_res =
+        unify_pat_types loc env ty_res expected_ty
+          ~refine:(refine || constr.cstr_generalized && no_existentials = None)
+      in
+      let expansion_scope = get_gadt_equations_level () in
+      let ty_args, ty_res =
+        match oty with
+          None ->
+            let ty_args, ty_res, _ =
+              instance_constructor ~in_pattern:(env, expansion_scope) constr in
+            unify_res ty_res;
+            ty_args, ty_res
+        | Some (nl, sty) ->
+            let ty_args, ty_res, ty_ex = instance_constructor constr in
+            unify_res ty_res;
+            let ids =
+              List.map
+                (fun name ->
+                  let decl = new_local_type ~expansion_scope ~loc:name.loc () in
+                  let (id, new_env) =
+                    Env.enter_type ~scope:expansion_scope name.txt decl !env in
```

Comment:
It's not immediately clear to me why this call is to `Env.enter_type` but the one inside of `Ctype.instance_constructor` is to `Env.add_local_type`. I think it would be good if we could find a way to factor these bits of code together.

Reply:
Good point. I have switched all creations of local types to `enter_type`, using only `add_local_type` for redefinitions.